net/netlink/netlink_attr.c

```diff
@@ -0,0 +1,328 @@
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <sys/types.h>
+#include <string.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+#include <unistd.h>
+
+#include <nuttx/net/netlink.h>
+
+#include "netlink.h"
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+#ifdef CONFIG_NETLINK_VALIDATE_POLICY
+static const uint8_t g_nla_attr_len[NLA_TYPE_MAX + 1] =
+{
+  0,
+  sizeof(uint8_t),
+  sizeof(uint16_t),
+  sizeof(uint32_t),
+  sizeof(uint64_t),
+  0, 0, 0, 0, 0, 0, 0,
+  sizeof(int8_t),
+  sizeof(int16_t),
+  sizeof(int32_t),
+  sizeof(int64_t),
+};
+
+static const uint8_t g_nla_attr_minlen[NLA_TYPE_MAX + 1] =
+{
+  0,
+  sizeof(uint8_t),
+  sizeof(uint16_t),
+  sizeof(uint32_t),
+  sizeof(uint64_t),
+  0, 0,
+  sizeof(uint64_t),
+  NLA_HDRLEN,
+  0, 0, 0,
+  sizeof(int8_t),
+  sizeof(int16_t),
+  sizeof(int32_t),
+  sizeof(int64_t),
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static int validate_nla_bitfield32(FAR const struct nlattr *nla,
+                                   FAR uint32_t *valid_flags_mask)
+{
+  FAR const struct nla_bitfield32 *bf = nla_data(nla);
+
+  if (!valid_flags_mask)
+    {
+      return -EINVAL;
+    }
+
+  /* disallow invalid bit selector */
+
+  if (bf->selector & ~*valid_flags_mask)
+    {
+      return -EINVAL;
+    }
+
+  /* disallow invalid bit values */
+
+  if (bf->value & ~*valid_flags_mask)
+    {
+      return -EINVAL;
+    }
+
+  /* disallow valid bit values that are not selected */
+
+  if (bf->value & ~bf->selector)
+    {
+      return -EINVAL;
+    }
+
+  return 0;
+}
+
+static int validate_nla(FAR const struct nlattr *nla, int maxtype,
+                        FAR const struct nla_policy *policy)
+{
+  FAR const struct nla_policy *pt;
+  int minlen = 0;
+  int attrlen = nla_len(nla);
+  int type = nla_type(nla);
+
+  if (type <= 0 || type > maxtype)
+    {
+      return -EINVAL;
+    }
+
+  pt = &policy[type];
+
+  DEBUGASSERT(pt->type <= NLA_TYPE_MAX);
+
+  if (g_nla_attr_len[pt->type] && attrlen != g_nla_attr_len[pt->type])
+    {
+      nwarn("netlink: '%d': attribute type %d has an invalid length.\n",
+            getpid(), type);
+      return -EINVAL;
+    }
+
+  switch (pt->type)
+    {
+      case NLA_FLAG:
+        if (attrlen > 0)
+          {
+            return -ERANGE;
+          }
+        break;
+
+      case NLA_BITFIELD32:
+        if (attrlen != sizeof(struct nla_bitfield32))
+          {
+            return -ERANGE;
+          }
+
+        return validate_nla_bitfield32(nla, pt->validation_data);
+
```

Comment:
Is this fall through as well (as NLA_NUL_STRING to NLA_STRING?), ok the return will avoid it to fall through, but for consistence makes sense to add the "break;" statement.